Terragrunt 101 (#6181)
gruntwork-io/terragrunt: docs/src/content/docs/02-guides/01-terragrunt-101/01-course-overview.md

---

> title: Course Overview

Reviewer: ```suggestion
title: Overview
```

> description: Welcome and orientation for the Terragrunt 101 self-paced course

Reviewer: ```suggestion
description: Welcome and orientation for the Terragrunt 101 self-paced guide
```

slug: guides/terragrunt-101
sidebar:
  order: 1
prev: false
---

## Welcome to Terragrunt 101

This is a **self-paced course** for engineers who already use OpenTofu or Terraform and want to level up their Infrastructure as Code (IaC) with Terragrunt.

---

### Why Terragrunt?

> Terragrunt helps you **organize**, **reuse**, and **orchestrate** infrastructure at scale, without the copy/paste and repetition that plain Terraform or OpenTofu often force.

Reviewer: ```suggestion
Terragrunt helps you **organize**, **reuse**, and **orchestrate** infrastructure at scale, without the copy/paste and repetition that OpenTofu or Terraform alone often require.
```

**What Terragrunt does really well today:**

> - **Stacks**: Split infrastructure into small, deployable **units** and combine them into **stacks**. Define a stack once (e.g., a VPC plus subnets plus security groups) and instantiate it across environments or accounts with different values. No duplicated directory trees.

Reviewer: ```suggestion
- **Stacks**: Split infrastructure into small, deployable [**units**](https://docs.terragrunt.com/getting-started/terminology/#unit) and combine them into [**stacks**](https://docs.terragrunt.com/getting-started/terminology/#stack). Define a stack once (e.g., a VPC plus subnets plus security groups) and instantiate it across environments or accounts with different values. No duplicated directory trees.
```

> - **Keep things DRY**: Share backend config, CLI arguments, and inputs across all your units via a single root config. The `backend` block doesn't support variables; Terragrunt does, so you write configuration once and reuse it everywhere.

Reviewer: ```suggestion
- **Keep things DRY**: Share backend config, CLI arguments, and inputs across all your units via a single root config. The OpenTofu/Terraform `backend` block doesn't support variables; Terragrunt does, so you write configuration once and reuse it everywhere.
```

> - **Catalog**: Scaffold new units quickly with `terragrunt catalog` from a curated set of modules instead of writing every `terragrunt.hcl` from scratch.

Reviewer: ```suggestion
- **Catalog**: Scaffold new units quickly with `terragrunt catalog` to enable infrastructure self-service using Infrastructure as Code.
```

> - **Orchestration**: Run many units in the right order. Terragrunt understands dependencies between modules and manages multiple state files so you can plan and apply in dependency order without manual scripting.

Reviewer: ```suggestion
- **Orchestration**: Run multiple units in the right order. Terragrunt understands dependencies between modules and manages multiple state files so you can plan, apply and destroy in dependency order without manual scripting.
```

> If you've outgrown ad hoc folder copying, repeated backend blocks, or custom scripts to chain Terraform/OpenTofu runs, this course will show you how Terragrunt addresses that. You'll learn the concepts (units, stacks, root config), the HCL blocks and functions that make config reusable, and how to author and compose stacks so your IaC stays maintainable as it grows.

Reviewer: ```suggestion
If you've outgrown ad-hoc copy/paste, repeated backend blocks, or custom scripts to chain OpenTofu/Terraform runs, this course will show you how Terragrunt addresses that. You'll learn the concepts (units, stacks, root config), the HCL blocks and functions that make config reusable, and how to author and compose stacks so your IaC stays maintainable as it grows.
```

---

#### A bit of history

Terragrunt came out of Gruntwork in July 2016. The team was running Terraform for dozens of clients and kept hitting the same limits: Terraform had no remote state locking, so if two people ran apply at once they could overwrite each other's state. The docs said locking wasn't supported and pointed you at Atlas (the paid product that became Terraform Enterprise). There was no way to express remote state as code, so every module got its backend config copy-pasted or hand-wired, which led to mistakes and duplicate resources.

> Terragrunt started as a thin wrapper that managed remote state for you and provided locking via DynamoDB. Over the next few years, Terraform adopted much of that (e.g., DynamoDB-style state locking landed in 0.9 in 2017). As those gaps closed, different pains became obvious: the `backend` block still didn't support variables, so teams kept copy-pasting. Common CLI flags had to be repeated on every run. Promoting the same module from dev to staging to prod meant copying whole directory trees. Coordinating multiple state files in the right order meant custom scripting. Terragrunt evolved to address those problems and kept evolving into what it is today: a tool built around keeping config DRY and helping teams split infrastructure into units and orchestrate them as stacks.

Reviewer: ```suggestion
Terragrunt started as a thin wrapper that managed remote state for you and provided locking via DynamoDB. Over the next few years, Terraform adopted some of that (e.g., DynamoDB-style state locking landed in Terraform v0.9 in 2017). As those gaps closed, different pains became obvious: the `backend` block still didn't support variables, so teams kept copy-pasting. Common CLI flags had to be repeated on every run. Promoting the same module from dev to staging to prod meant copying whole directory trees. Coordinating multiple state files in the right order meant custom scripting. Terragrunt evolved to address those problems and kept evolving into what it is today: an orchestration tool built around splitting infrastructure into stacks of units that could be run independently in a coordinated manner, and helping teams keep their config DRY.
```

## What You'll Learn

Terragrunt helps you **organize** and **reuse** your infrastructure code.

---

### Getting Started

You'll start with the basics in [**Module 2: First Principles**](/guides/terragrunt-101/first-principles-terminology/), where you'll learn:

> - What are **units** and **stacks**?

Reviewer: ```suggestion
- What **units** and **stacks** are
```

Also, please add a space above this.

> - How does Terragrunt run your OpenTofu or Terraform code?

Reviewer: ```suggestion
- How Terragrunt runs OpenTofu/Terraform code
```

Then in [**Module 3: Getting Started**](/guides/terragrunt-101/getting-started/), you'll install Terragrunt, learn the commands, and set up your first project.

---

### Building Skills

[**Module 4: HCL Blocks**](/guides/terragrunt-101/blocks/) introduces the `terraform`, `include`, and `dependency` blocks that power Terragrunt configurations.

> [**Module 5: Functions**](/guides/terragrunt-101/functions/) covers path helpers, environment lookups, and shell execution—tools that help you share configuration across environments *without copying and pasting*.

Reviewer: ```suggestion
[**Module 5: Functions**](/guides/terragrunt-101/functions/) covers path helpers, environment lookups, and shell execution: tools that help you share configuration across environments *without copying and pasting*.
```

In [**Module 6: Authoring Units**](/guides/terragrunt-101/authoring-units/), you'll write your own reusable units using inputs, generate blocks, and include patterns.

> [**Module 7: Composing Stacks**](/guides/terragrunt-101/composing-stacks/) teaches you to combine units into stacks with dependencies—both implicit and explicit.

Reviewer: ```suggestion
[**Module 7: Composing Stacks**](/guides/terragrunt-101/composing-stacks/) teaches you to combine units into stacks with dependencies.
```

---

### Advanced Topics

[**Module 8: Advanced Patterns**](/guides/terragrunt-101/advanced-patterns/) brings it all together with:

> - Dynamic credentials with `auth-provider-command`

Reviewer: Space above this, please.

> - Run-queue strategies and filtering

Reviewer: ```suggestion
- Run queue strategies and filtering
```

- Error handling (retries and ignoring expected errors)

## Resources

### Terragrunt Documentation

For command references and configuration details:

| Resource | Description |

> |:---------|:------------|

Reviewer: Please address the markdownlint findings here.

> | [**Full documentation**](https://terragrunt.gruntwork.io/docs/) | Complete reference |

Reviewer: ```suggestion
| [**Full documentation**](/docs/) | Complete reference |
```

> | [**Configuration blocks**](https://terragrunt.gruntwork.io/docs/reference/hcl/blocks/) | HCL block reference |

Reviewer: ```suggestion
| [**Configuration blocks**](/hcl/blocks/) | HCL block reference |
```

> | [**Built-in functions**](https://terragrunt.gruntwork.io/docs/reference/hcl/functions/) | Function reference |

Reviewer: ```suggestion
| [**Built-in functions**](/docs/reference/hcl/functions/) | Function reference |
```

> | [**CLI reference**](https://terragrunt.gruntwork.io/docs/reference/cli/) | Command-line options |

Reviewer: ```suggestion
| [**CLI reference**](/docs/reference/cli/) | Command-line options |
```
